### PR TITLE
Implement NSMeasurement

### PR DESCRIPTION
**Pull request by Dante-Broggi**
This is implementing the `NSUnimplemented` methods.

### CODE REVIEW

**Comment by ianpartridge** [comment]
Thanks for your contribution! Can you add some tests too?

P.S. We use 4 spaces for indentation here.

File: Foundation/NSMeasurement.swift [hunk 1]
```diff
@@ -23,21 +23,72 @@ open class NSMeasurement : NSObject, NSCopying, NSSecureCoding {
...
+    open func copy(with zone: NSZone? = nil) -> Any { return self }
     
     open class var supportsSecureCoding: Bool { return true }
     
-    open func encode(with aCoder: NSCoder) { NSUnimplemented() }
+    open func encode(with aCoder: NSCoder) { 
+      guard aCoder.allowsKeyedCoding else {
+            preconditionFailure("Unkeyed coding is unsupported.")
+        }
+        aCoder.encode(self.doubleValue, forKey:"NS.dblval")
```
Comment (parkera) [inline Foundation/NSMeasurement.swift]: `NS.value`

**Comment by Dante-Broggi** [comment]
For tests, as there isn't a `TestNSMeasurement.swift` file, do I have to do anything special when adding it?

**Comment by millenomi** [comment]
Please add it to the Xcode project and the list of test source files in the build.py file when you do.

**Comment by Dante-Broggi** [comment]
The `build.py` includes `+ glob.glob('./TestFoundation/Test*.swift')) # all TestSomething.swift are considered sources to the test project in the TestFoundation directory` so, actually I don't need to add `TestNSMeasurement.swift`, but I did add it to the Xcode project.

**Comment by spevans** [comment]
@swift-ci please test

**Comment by spevans** [comment]
Looks like there were a couple of compilation errors, see https://ci.swift.org/view/Pull%20Request/job/swift-corelibs-foundation-PR-Linux/635/consoleFull#18229161593122a513-f36a-4c87-8ed7-cbc36a1ec144

**Comment by spevans** [comment]
@swift-ci please test

**Comment by spevans** [comment]
@Dante-Broggi there are still a couple of compile errors in CI: https://ci.swift.org/view/Pull%20Request/job/swift-corelibs-foundation-PR-Linux/636/console

**Comment by Dante-Broggi** [comment]
So, the one thing I couldn't fix is `canBeConverted(to:)`, the original implementation of which used objc-only `isKind(of:)` and I cannot think of how to replace that.
The nearest I can think of is `(otherUnit is type(of: unit))`, except `is` would require a dynamic metatype argument on the rhs, and it currently only accepts static metatypes.

**Comment by Dante-Broggi** [comment]
I thought of something that doesn't use `.isKind(of:)` for `canBeConverted(to:)`
*just attempt the conversion*, and return false instead of `fatalError()`ing.

**Comment by spevans** [comment]
@swift-ci please test

File: Foundation/NSMeasurement.swift [hunk 2]
```diff
@@ -23,21 +23,95 @@ open class NSMeasurement : NSObject, NSCopying, NSSecureCoding {
...
+        }
+    }
     
-    open func adding(_ measurement: Measurement<Unit>) -> Measurement<Unit> { NSUnimplemented() }
+    open func adding(_ rhs: Measurement<Unit>) -> Measurement<Unit> {
+        if self.unit.isEqual(rhs.unit) {
+            return Measurement(value: self.doubleValue + rhs.value, unit: self.unit)
+        } else {
+            guard let dimension = unit as? Dimension,
+                    let otherDimension = otherUnit as? Dimension else {
```
Comment (spevans) [inline Foundation/NSMeasurement.swift]: Do you mean `rhs` instead of `otherUnit` ? (same on line 82)

File: Foundation/NSMeasurement.swift [hunk 3]
```diff
@@ -23,21 +23,95 @@ open class NSMeasurement : NSObject, NSCopying, NSSecureCoding {
...
+            return otherUnit.isKind(of: type(of: unit))
+        #else
+            // just check conversion
+            if let dimension = unit as? Dimension,
+               let otherDimension = otherUnit as? Dimension {
+                return true
+            } else {
+                return unit.isEqual(otherUnit)
+            }
+            re
```
Comment (spevans) [inline Foundation/NSMeasurement.swift]: Extra typo here

File: Foundation/NSMeasurement.swift (hunk 2 above)
Comment (Dante-Broggi) [reply Foundation/NSMeasurement.swift]: Yes, indeed.

**Comment by spevans** [comment]
@swift-ci please test

**Comment by spevans** [comment]
@swift-ci please test

File: Foundation/NSMeasurement.swift [hunk 4]
```diff
@@ -23,21 +23,94 @@ open class NSMeasurement : NSObject, NSCopying, NSSecureCoding {
...
     
-    open func adding(_ measurement: Measurement<Unit>) -> Measurement<Unit> { NSUnimplemented() }
+    open func adding(_ rhs: Measurement<Unit>) -> Measurement<Unit> {
+        if self.unit.isEqual(rhs.unit) {
+            return Measurement(value: self.doubleValue + rhs.value, unit: self.unit)
+        } else {
+            guard let dimension = unit as? Dimension,
+                    let otherDimension = rhs.unit as? Dimension else {
+                        fatalError("Cannot convert differing units that are non-dimensional! lhs: \(type(of: dimension)) rhs: \(type(of: otherDimension))")
+                }
```
Comment (spevans) [inline Foundation/NSMeasurement.swift]: This needs to be `(type(of: unit))` and `(type(of: rhs.unit))` as `dimension` and `otherDimension` don't exist in th `guard` block if the `as?` fails.

btw, are you testing on macOS or Linux?

File: Foundation/NSMeasurement.swift [hunk 5]
```diff
@@ -23,21 +23,94 @@ open class NSMeasurement : NSObject, NSCopying, NSSecureCoding {
...
     
-    open func canBeConverted(to unit: Unit) -> Bool { NSUnimplemented() }
+    open func canBeConverted(to otherUnit: Unit) -> Bool {
+        #if DEPLOYMENT_RUNTIME_OBJC //|| os(Linux)
+            return otherUnit.isKind(of: type(of: unit))
+        #else
+            // just check conversion
+            if let dimension = unit as? Dimension,
+               let otherDimension = otherUnit as? Dimension {
+                return true
```
Comment (spevans) [inline Foundation/NSMeasurement.swift]: `dimension` and `otherDimension` are not used so could be replaced with `_` or perhaps rewrite as `unit is Dimension && otherUnit is Dimension`.

Comment (Dante-Broggi) [reply Foundation/NSMeasurement.swift]: I think I thought of that, then forgot to do it.

File: Foundation/NSMeasurement.swift (hunk 4 above)
Comment (Dante-Broggi) [reply Foundation/NSMeasurement.swift]: I am on a mac, but I really have no idea how to actually even build any of the swift projects.
So, realistically, nowhere. I am sorry, but I am just an enthusiast with no idea how to use the infrastructure. 🙁

**Comment by spevans** [comment]
@Dante-Broggi Ok here's what to do assuming you are on either Sierra (10.12) or High Sierra (10.13) and have Xcode already installed:

Get the latest snapshot from https://swift.org/download/#snapshots Trunk development (master) 'Xcode' . This is a package you install on your mac.

Next, in Xcode, goto Xcode > Toolchains and you should see the snapshot you just downloaded in the menu, select it as your default tool chain. It will be called something like 'Swift Development Snapshot 2018-02-08 (a)' or something like that.

Now in the same directory that you checked out the `swift-corelibs-foundation` repo you also need to checkout the `swift-corelibs-xctest` repo.

Now in the `swift-corelibs-foundation` directory, open the `Foundation.xcworkspace` workspace (NOT the xcodeproj). The workspace includes the Foundation Xcode project and the XCTest Xcode project.

In the menu at the top the Xcode window, select 'TestFoundation > My Mac' then do 'Product > Run' and that will compile up CodeFoundation, Swift Foundation, XCtest and TestFoundation and then it will try and run the tests. Any compile errors should show up here. Don't worry about all of the warnings about deprecation in code that you haven't changed, there is usually because of changes in the main swift core in the snapshot you downloaded.

That should be enough to get you started, let me know if you have any questions about the above.

**Comment by spevans** [comment]
@swift-ci please test

File: Foundation/NSMeasurement.swift [hunk 6]
```diff
@@ -23,21 +23,93 @@ open class NSMeasurement : NSObject, NSCopying, NSSecureCoding {
         self.unit = unit
     }
     
-    open func canBeConverted(to unit: Unit) -> Bool { NSUnimplemented() }
+    open func canBeConverted(to otherUnit: Unit) -> Bool {
+        #if DEPLOYMENT_RUNTIME_OBJC //|| os(Linux)
```
Comment (xwu) [inline Foundation/NSMeasurement.swift]: Nit: the stuff inside a compilation conditional should be indented as though there weren't a conditional surrounding it; in most files, the conditional statement itself should be outdented to the left margin, but the specific style can vary depending on the file.

File: Foundation/NSMeasurement.swift [hunk 7]
```diff
@@ -23,21 +23,93 @@ open class NSMeasurement : NSObject, NSCopying, NSSecureCoding {
...
+            return otherUnit.isKind(of: type(of: unit))
+        #else
+            // just check conversion
+            if unit is Dimension && otherUnit is Dimension {
+                return true
+            } else {
+                return unit.isEqual(otherUnit)
+            }
+        #endif
```
Comment (xwu) [inline Foundation/NSMeasurement.swift]: Does this implementation match Darwin Foundation? It seems to be very limited.

File: Foundation/NSMeasurement.swift [hunk 8]
```diff
@@ -23,21 +23,93 @@ open class NSMeasurement : NSObject, NSCopying, NSSecureCoding {
...
     
-    open func converting(to unit: Unit) -> Measurement<Unit> { NSUnimplemented() }
+    open func converting(to otherUnit: Unit) -> Measurement<Unit> {
+        if canBeConverted(to: otherUnit) {
+            if unit.isEqual(otherUnit) {
+                return Measurement(value: doubleValue, unit: otherUnit)
+            } else {
+                guard let dimension = unit as? Dimension,
+                    let otherDimension = otherUnit as? Dimension else {
+                        fatalError("Cannot convert differing units that are non-dimensional! lhs: \(type(of: unit)) rhs: \(type(of: otherUnit))")
```
Comment (xwu) [inline Foundation/NSMeasurement.swift]: Try to align these error messages as closely as possible with Darwin Foundation.

File: Foundation/NSMeasurement.swift [hunk 9]
```diff
@@ -23,21 +23,93 @@ open class NSMeasurement : NSObject, NSCopying, NSSecureCoding {
...
+    }
     
-    open func converting(to unit: Unit) -> Measurement<Unit> { NSUnimplemented() }
+    open func converting(to otherUnit: Unit) -> Measurement<Unit> {
+        if canBeConverted(to: otherUnit) {
+            if unit.isEqual(otherUnit) {
+                return Measurement(value: doubleValue, unit: otherUnit)
+            } else {
+                guard let dimension = unit as? Dimension,
+                    let otherDimension = otherUnit as? Dimension else {
```
Comment (xwu) [inline Foundation/NSMeasurement.swift]: Nit: the stuff inside the braces should be indented only four spaces as compared to `guard`; it's fine to indent the second line of the condition but it doesn't affect how the inside stuff is indented (the same applies elsewhere below).

File: Foundation/NSMeasurement.swift [hunk 10]
```diff
@@ -23,21 +23,93 @@ open class NSMeasurement : NSObject, NSCopying, NSSecureCoding {
...
+    open func copy(with zone: NSZone? = nil) -> Any { return self }
     
     open class var supportsSecureCoding: Bool { return true }
     
-    open func encode(with aCoder: NSCoder) { NSUnimplemented() }
+    open func encode(with aCoder: NSCoder) {
+        guard aCoder.allowsKeyedCoding else {
+            preconditionFailure("Unkeyed coding is unsupported.")
+        }
+        aCoder.encode(self.doubleValue, forKey:"NS.value")
```
Comment (xwu) [inline Foundation/NSMeasurement.swift]: Nit: space after `forKey:`; likewise one line below.

File: Foundation/NSMeasurement.swift [hunk 11]
```diff
@@ -23,21 +23,93 @@ open class NSMeasurement : NSObject, NSCopying, NSSecureCoding {
...
     
-    public required init?(coder aDecoder: NSCoder) { NSUnimplemented() }
+    public required init?(coder aDecoder: NSCoder) {
+        guard aDecoder.allowsKeyedCoding else {
+            preconditionFailure("Unkeyed coding is unsupported.")
+        }
+        let doubleValue = aDecoder.decodeDouble(forKey: "NS.value")
+        let possibleUnit = aDecoder.decodeObject(forKey: "NS.unit")
+        guard let unit = possibleUnit as? Unit else {
+            return nil // or should we `fatalError()`?
```
Comment (xwu) [inline Foundation/NSMeasurement.swift]: What does Darwin Foundation do?

Comment (malhal) [reply Foundation/NSMeasurement.swift]: One thing it does is validation, e.g. @"Cannot add measurements of differing unit types!"

**Comment by alblue** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
@Dante-Broggi are you still interested in chaperoning this patch?

**Comment by ikesyo** [comment]
Looks like this is superseded by #2095.

**Pull request closed without merging**